File: tests/qt_types_standalone/rust/src/qmodelindex.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QModelIndex;

#[cxx::bridge]
mod qmodelindex_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qmodelindex.h");
        type QModelIndex = cxx_qt_lib::QModelIndex;
    }

    #[namespace = "rust::cxxqtlib1"]
    unsafe extern "C++" {
        include!("cxx-qt-lib/common.h");
        type c_void = cxx_qt_lib::c_void;
    }

    extern "Rust" {
        fn construct_qmodelindex() -> QModelIndex;
        fn read_qmodelindex(i: &QModelIndex) -> bool;
        fn clone_qmodelindex(i: &QModelIndex) -> QModelIndex;
        fn internal_pointer_qmodelindex(i: &QModelIndex) -> *mut c_void;
        fn internal_id_qmodelindex(i: &QModelIndex) -> usize;
    }
}

fn construct_qmodelindex() -> QModelIndex {
    QModelIndex::default()
}

fn read_qmodelindex(i: &QModelIndex) -> bool {
    i.is_valid() && i.row() == 0
}

fn clone_qmodelindex(i: &QModelIndex) -> QModelIndex {
    i.clone()
}

fn internal_pointer_qmodelindex(i: &QModelIndex) -> *mut qmodelindex_cxx::c_void {
    i.internal_pointer_mut()
}

fn internal_id_qmodelindex(i: &QModelIndex) -> usize {
    i.internal_id()
}

File: crates/cxx-qt-lib/src/core/qmodelindex.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qmodelindex.h"

#include <cxx-qt-lib/assertion_utils.h>

// QModelIndex has two ints, a quint pointer (same as size_t), and a pointer.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/itemmodels/qabstractitemmodel.h?h=v5.15.6-lts-lgpl#n93
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/itemmodels/qabstractitemmodel.h?h=v6.2.4#n195
assert_alignment_and_size(QModelIndex, {
  ::std::int32_t a0;
  ::std::int32_t a1;
  ::std::size_t a2;
  ::std::size_t a3;
});

static_assert(::std::is_trivially_copyable<QModelIndex>::value);
